crates/lang/macro/src/lib.rs

```diff
@@ -18,20 +18,600 @@ mod trait_def;
+///
+/// If you are a beginner trying to learn ink! we recommend you to check out
+/// our extensive [ink! workshop](https://substrate.dev/substrate-contracts-workshop/#/).
+///
+/// **Note:** In all below examples we will be using `ink_lang` crate aliased as just `ink`.
+///           You can do this yourself by adding the following line to your code:
+///           `use ink_lang as ink;`
+///
+/// # Description
+///
+/// The macro does analysis on the provided smart contract code and generates
+/// proper code.
+///
+/// ink! smart contracts can compile in several different modes.
+/// There are two main compilation models using either
+/// - on-chain mode: `no_std` + WebAssembly as target
+/// - off-chain mode: `std`
+///
+/// We generally use the on-chain mode for actual smart contract deployment
+/// whereas we use the off-chain mode for smart contract testing using the
+/// off-chain environment provided by the `ink_env` crate.
+///
+/// # Usage
+///
+/// ## Header Arguments
+///
+/// The `#[ink::contract]` macro can be provided with some additional comma-separated
+/// header arguments:
+///
+/// - `dynamic_storage_allocator: bool`
+///
+///     Tells the ink! code generator to allow usage of ink!'s built-in dynamic
+///     storage allocator.
+///     - `true`: Use the dynamic storage allocator provided by ink!.
+///     - `false`: Do NOT use the dynamic storage allocator provided by ink!.
+///
+///     This feature is generally only needed for smart contracts that try to model
+///     their data in a way that contains storage entites within other storage
+///     entities.
+///     An example for this is the following type that could potentially be used
+///     within an ink!'s storage struct definition:
+///     ```
+///     // A storage vector of storage vectors.
+///     # use ink_storage as storage;
+///     # type _unused =
+///     storage::Vec<storage::Vec<i32>>
+///     # ;
+///     ```
+///
+///     **Usage Example:**
+///     ```
+///     # use ink_lang as ink;
+///     #[ink::contract(dynamic_storage_allocator = true)]
+///     mod my_contract {
+///         # #[ink(storage)]
+///         # pub struct MyStorage;
+///         # impl MyStorage {
+///         #     #[ink(constructor)]
+///         #     pub fn construct() -> Self { MyStorage {} }
+///         #     #[ink(message)]
+///         #     pub fn message(&self) {}
+///         # }
+///         // ...
+///     }
+///     ```
+///
+///     **Default value:** `false`
+///
+/// - `compile_as_dependency: bool`
+///
+///     Tells the ink! code generator to always or never
+///     compile the smart contract as if it was used as a dependency of another ink!
+///     smart contract.
+///     Normally this flag is only really useful for ink! developers of people that
+///     want to inspect code generation of ink! smart contracts.
+///     The author is not aware of any particular practical use case for users that
+///     makes use of this flag.
```

Comment:
can you make a suggestions? :sweat_smile: Or tell me what is unclear to you?

Reply:
Isn't this flag used for when cross-calling contracts, the callee contract is compiled as a dependency? Although is it intended that traits should be used in this case?

Anyway it isn't clear from this comment what the flag is actually for, only that there is no "practical use case". As a user I would not be further enlightened.

Reply:
Well maybe there are use cases that we don't know yet but normally you compile ink! smart contracts as dependency through their crate features `ink-as-dependency` and not via this flag because you do not necessarily always want to compile them as dependency. Or maybe you do? I want to leave this question open for further research. After all using ink! smart contracts yields many discoveries.